Add option to ignore colorspace information in texture images

j9liu opened the pull request:
Fixes #9390.

The `UNPACK_COLORSPACE_CONVERSION_WEBGL` flag was not set to `NONE` as specified in the glTF [documentation](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#texture-data).

When the Sandcastle from the issue is copy pasted into a local session, it should now look like this. The spheres across each row are identical.

![image](https://user-images.githubusercontent.com/32226860/122474217-ee964d80-cf90-11eb-8f6c-2dfa1a6da3bb.png)

cc @ebogo1 @shehzan10

cesium-concierge commented:
Thanks for the pull request @j9liu!
* :heavy_check_mark: Signed CLA found.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

lilleyse commented:
The change seems good for glTF but it could impact rendering of non-glTF images such as imagery on terrain, textured polygons, or billboards. You might notice something if you compare sandcastle examples on this branch vs. `master` (or maybe not).

The safe thing to do would be to add an option to `Texture` called `skipColorSpaceConversion` that only `Model.js` sets to true.

CC @ptrgags @sanjeetsuhag for the `Model.js` rewrite

j9liu commented:
@lilleyse - just pushed some changes, is this better?

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -2958,6 +2958,7 @@ function createTexture(gltfTexture, model, context) {
       pixelDatatype: texture.type,
       sampler: sampler,
       flipY: false,
+      skipColorSpaceConversion: true,
```
I forgot to mention this earlier - can you add this option to `GltfTextureLoader` as well? That file is part of the `Model.js` rewrite that @ptrgags and @sanjeetsuhag are working on but we might as well make the fix there now. See https://github.com/CesiumGS/cesium/blob/master/Source/Scene/GltfTextureLoader.js#L272

lilleyse commented on Source/Renderer/Texture.js:
```diff
@@ -197,6 +197,15 @@ function Texture(options) {
 
   gl.pixelStorei(gl.UNPACK_ALIGNMENT, unpackAlignment);
 
+  if (options.skipColorSpaceConversion) {
+    gl.pixelStorei(gl.UNPACK_COLORSPACE_CONVERSION_WEBGL, gl.NONE);
+  } else {
+    gl.pixelStorei(
+      gl.UNPACK_COLORSPACE_CONVERSION_WEBGL,
+      gl.BROWSER_DEFAULT_WEBGL
+    );
+  }
```
This same code should also go in `Texture.prototype.copyFrom`

j9liu replied:
Just to clarify, in this function, I should change `options` to `source`, right?

lilleyse commented on Source/Renderer/Texture.js:
```diff
@@ -197,6 +197,15 @@ function Texture(options) {
   gl.pixelStorei(gl.UNPACK_ALIGNMENT, unpackAlignment);
 
+  if (options.skipColorSpaceConversion) {
+    gl.pixelStorei(gl.UNPACK_COLORSPACE_CONVERSION_WEBGL, gl.NONE);
```
There are a few cases where this fix alone is not enough.

It works when `source` is an `HTMLImageElement` - this happens when `fetchImage` is called on an external image and `preferImageBitmap` is false or `ImageBitmap` is not supported. Basically glTFs with external images are ok for now...

But embedded images won't work because they use `ImageBitmap` by default (via `loadImageFromTypedArray`) and according to the specs `UNPACK_COLORSPACE_CONVERSION_WEBGL` is ignored for `ImageBitmap`. See https://www.khronos.org/registry/webgl/specs/latest/1.0/#6.10

Another problematic case is when non-power-of-two images are rescaled on a canvas. `HTMLCanvasElement` doesn't seem to get affected by the `UNPACK_COLORSPACE_CONVERSION_WEBGL` flag either. That happens when a texture has to be rescaled to meet WebGL 1's power-of-two requirements for certain samplers. If you simulate that by removing the if guard around [this code](https://github.com/CesiumGS/cesium/blob/3ca010a77a093b2c6b472e0fe307a7ce5cef18db/Source/Scene/Model.js#L2934-L2952) you'll see the problems show up again.

`ImageBitmap` may be fixable according to the specs. I haven't looked into `HTMLCanvasElement` but it's worth investigating.

![image](https://user-images.githubusercontent.com/915398/122625762-c2ec9380-d074-11eb-8548-ea80ffb6ef72.png)

lilleyse replied:
CC https://github.com/CesiumGS/cesium/pull/9627. After that PR is merged this fix won't work for external images either since they will now be `ImageBitmap`.

lilleyse replied:
For testing embedded images use the glb version of the model: https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/TextureEncodingTest/glTF-Binary

j9liu replied:
In the [Sandcastle](https://sandcastle.cesium.com/#c=dVTvT9swEP1XrH4hSJmTjp8tBQ0B0yYNjbGKL8s+mOTSWHPsyj4Xuqn/+85J04YC/uTcvXv37tmOAmQI1gqp2Tm7Aid9zXMLAuEaSuEVTtvsnTULWYB9kPB0awpQLto/y3SmFTEsKAiWCDQ8dSQPTSzay5vPK6ORWMDuxexfphktBwpylEZ/1YXMBRrLxqwUykHcAvC9zuMuFfeZoHhP6gb/a/i7K6mMV8WlljVNSnnrIdOrdqLS60YXa31oKCJvVcwqkLMK97sJ2rE5aJQowXELtVnApVJrawIm2DM3TjaEG4evhEXaCX3AS2vqa5hZABcND0b8eDQ8SY9jdnDEj09Hh6cnm7ZnW8YKRCH1bEt4K7DiaO4pLLSLRmkfPZeYV4RN+VEvao1SIbgO9cjnO0f5pe12F2juqSpat49b5rih6jc0VpIp4uXMU0vKSmNrx6sdwh+efLaa4FHnVRxkvLSx8XlJhLu+i6KI1kcS1sbu8WYbb7N9beP+Vw9ThyOnbI80LG8lBcNNeBl3uVBAmWGapjzd5lbtdtVZs9ZdquXURO009BpY4W0jYMwOCcuxAh11dzDa74tYE6AV+R8objo/WqqzXrdV9zJJLAGyQYU4d+MkyZ8/pnwmsfKPXJpkprD8QDcRE/T0AKVQ7QtLpvCM3sKNzk04qWmAzNT081sJHliyQXNWr57MEXkSFA3iwcThUsFFN84nWc+NbSRGnCcI9VxRrUsePU2HPHeuM26S9EsnhVwwWZxng51/SzZguRLOUab0Sv2UfyEbXEwSwr8qVaa5gt8XYJVYBlg1vPjWBjnnk4Q+365EY9SjsDvM/wE) I'm using to test this, how do I test the glb version of the model? If I replace the glTF URL with the .glb one, it gives me an error, even if I add the `?raw=true` flag at the end of the URL to get the raw file.

j9liu replied:
Also, it seems that there is a flag in `ImageBitmapOptions` that is equivalent to `UNPACK_COLORSPACE_CONVERSION_WEBGL`, but it will require me modifying functions like `createImageBitmapFromBlob` and `loadImageFromTypedArray`, is that ok?

lilleyse replied:
> it will require me modifying functions like createImageBitmapFromBlob and loadImageFromTypedArray, is that ok?

Yup that's the right approach. I left a comment with more detail about that.

lilleyse replied:
Curious, did you have a chance to look into into the `HTMLCanvasElement` problem?

lilleyse replied:
Never mind, I saw your comment below.

j9liu replied:
Hm, I'm unable to reproduce the HTMLCanvasElement problem. I removed the `if` guards around the code that you linked to, but my result was the same whether or not they were there.

j9liu replied:
I tried to look into this anyway, and unfortunately, the image handling functions in [`CanvasContextRendering2D`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D) have no options for handling color profiles. There's a [proposal](https://wiki.whatwg.org/wiki/CanvasColorSpace) to add a feature that specifies a color profile for a context, but nothing actually official.

lilleyse replied:
Ok thanks for looking into it. This is a less common case so I'm ok with moving forward with the PR.

j9liu commented:
@lilleyse - what do you think of these changes? I haven't found a workaround for the `HTMLCanvasElement` situation yet, but I'll keep looking.

lilleyse commented on Source/Core/loadImageFromTypedArray.js:
```diff
@@ -12,6 +12,10 @@ function loadImageFromTypedArray(options) {
   var request = options.request;
   var flipY = defaultValue(options.flipY, false);
+  var skipColorSpaceConversion = defaultValue(
+    options.skipColorSpaceConversion,
```
Tests should be added for this new option. Does the `ImageBitmap` store the `colorSpaceConversion` it was created with? Hopefully, because that will make it pretty straightforward to test.

lilleyse replied:
Same with the `Resource` functions.

lilleyse commented on Source/Renderer/Texture.js:
```diff
@@ -347,6 +360,7 @@ function Texture(options) {
   this._sizeInBytes = sizeInBytes;
   this._preMultiplyAlpha = preMultiplyAlpha;
   this._flipY = flipY;
+  this._skipColorSpaceConversion = skipColorSpaceConversion;
```
I don't think this should to be saved as a member variable of `Texture`. Instead it should be an option in `Texture` and `Texture.prototype.copyFrom`.

The way I think of it is it's not the texture that inherently says to skip the color conversion, it's an option that's tied to the source data being sent through.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -2779,6 +2779,7 @@ function loadTexturesFromBufferViews(model) {
         uint8Array: loadResources.getBuffer(bufferView),
         format: gltfTexture.mimeType,
         flipY: false,
+        skipColorSpaceConversion: true,
```
There's also a `fetchImage` call the option should be passed to.

lilleyse commented on Source/Scene/GltfImageLoader.js:
```diff
@@ -259,6 +259,7 @@ function loadImageFromBufferTypedArray(typedArray) {
     uint8Array: typedArray,
     format: mimeType,
     flipY: false,
+    skipColorSpaceConversion: true,
```
Add the option to the `fetchImage` call in here as well.

lilleyse commented on Source/Core/Resource.js:
```diff
@@ -925,6 +926,7 @@ Resource.prototype.fetchImage = function (options) {
         return Resource.createImageBitmapFromBlob(blob, {
           flipY: flipY,
           premultiplyAlpha: false,
+          skipColorSpaceConversion: true,
```
`skipColorSpaceConversion: true` probably shouldn't be hardcoded anywhere in `Resource.js`. The option should be added to `Resource.prototype.fetchImage` and `Resource.fetchImage` and default to false. Since these functions are part of the public API the new option should be mentioned in `CHANGES.md`. The JSDoc should also be updated to reflect these new properties.

Then make sure to thread the option through the rest of the image loading functions.

lilleyse commented on Source/Renderer/Texture.js:
```diff
@@ -624,12 +638,16 @@ Object.defineProperties(Texture.prototype, {
-Texture.prototype.copyFrom = function (source, xOffset, yOffset) {
-  xOffset = defaultValue(xOffset, 0);
-  yOffset = defaultValue(yOffset, 0);
+Texture.prototype.copyFrom = function (options) {
```
Make sure to update the example in the documentation above.

lilleyse commented on Specs/Renderer/TextureSpec.js:
```diff
@@ -731,7 +735,11 @@ describe(
+      });
 
       // Now green on top
       txCoords = new Cartesian2(0.5, 0.75);
```
The tests below that call `copyFrom` also need to be updated.

lilleyse commented on Source/Core/Resource.js:
```diff
@@ -1000,6 +1009,7 @@ function fetchImage(options) {
       crossOrigin,
       deferred,
       flipY,
+      skipColorSpaceConversion,
```
I did a global search for `createImage` and noticed that some terrain and imagery tests require updates. See `ArcGisMapServerImageryProviderSpec.js` and others examples like below - the new parameter needs to be added.

```
          Resource._DefaultImplementations.createImage(
            request,
            crossOrigin,
            deferred,
            true,
            true
          );
```

lilleyse commented on Specs/Core/loadImageFromTypedArraySpec.js:
```diff
@@ -62,6 +63,50 @@ describe("Core/loadImageFromTypedArray", function () {
+          expect(window.createImageBitmap).toHaveBeenCalledWith(blob, {
+            imageOrientation: "none",
+            premultiplyAlpha: "none",
+            colorSpaceConversion: "default",
```
Excellent tests throughout!

lilleyse commented on Source/Renderer/Texture.js:
```diff
@@ -603,11 +616,12 @@ Object.defineProperties(Texture.prototype, {
  * or an object with width, height, and arrayBufferView properties.
- *
- * @param {Object} source The source {@link ImageData}, {@link HTMLImageElement}, {@link HTMLCanvasElement}, or {@link HTMLVideoElement},
+ * @param {Object} options Object with the following properties:
```
For the sake of API symmetry we should also make the same changes to `CubeMapFace.prototype.copyFrom`.

lilleyse commented:
@j9liu thanks for the updates, they look really good overall. Just a few small things left and this PR should be ready to go.

j9liu commented:
@lilleyse - I think I addressed all your comments; let me know if I missed anything!

Also, I noticed that these two tests fail when I run them -- should I look into them? They were failing before I made the major changes to `Texture` and `Resource`, so I'm not sure if it is related to what I did.

![image](https://user-images.githubusercontent.com/32226860/123160414-2a6f5e00-d43c-11eb-974f-fe8c5347f2e2.png)

lilleyse commented on Source/Renderer/loadCubeMap.js:
```diff
@@ -13,6 +13,7 @@ import CubeMap from "./CubeMap.js";
  *
  * @param {Context} context The context to use to create the cube map.
  * @param {Object} urls The source URL of each image.  See the example below.
+ * @param {Boolean} skipColorSpaceConversion If true, any custom gamma or color profiles in the images will be ignored.
```
```suggestion
 * @param {Boolean} [skipColorSpaceConversion=false] If true, any custom gamma or color profiles in the images will be ignored.
```

lilleyse commented on Specs/Scene/ArcGisMapServerImageryProviderSpec.js:
```diff
@@ -224,7 +224,9 @@ describe("Scene/ArcGisMapServerImageryProvider", function () {
           Resource._DefaultImplementations.createImage(
             request,
             crossOrigin,
-            deferred
+            deferred,
+            true,
+            true
```
Ah I should have been clearer about this. The only functions that need to be updated are those that were already passing in the optional parameters like [here](https://github.com/CesiumGS/cesium/blob/d9ab862e4095ecd960555bd60a05f984c378fe23/Specs/Scene/ArcGisMapServerImageryProviderSpec.js#L337-L343). Since `createImage` now takes 6 parameters code like below needs to be changed


```
          Resource._DefaultImplementations.createImage(
            request,
            crossOrigin,
            deferred,
            true,
            true
          );
```
changes to

```
          Resource._DefaultImplementations.createImage(
            request,
            crossOrigin,
            deferred,
            true,
            false,
            true,
          );
```

lilleyse commented on Source/Scene/SkyBox.js:
```diff
@@ -146,7 +146,7 @@ SkyBox.prototype.update = function (frameState, useHdr) {
     if (typeof sources.positiveX === "string") {
       // Given urls for cube-map images.  Load them.
-      loadCubeMap(context, this._sources).then(function (cubeMap) {
+      loadCubeMap(context, this._sources, false).then(function (cubeMap) {
```
Fine to use the default value here

```suggestion
      loadCubeMap(context, this._sources).then(function (cubeMap) {
```

lilleyse commented:
> Also, I noticed that these two tests fail when I run them -- should I look into them? They were failing before I made the major changes to `Texture` and `Resource`, so I'm not sure if it is related to what I did.

It looks like the tests pass in `master` so I think it's a regression in this PR. There's a good chance that once https://github.com/CesiumGS/cesium/pull/9624#discussion_r657499492 is addressed they will be fixed.

j9liu commented:
@lilleyse - sorry for the misunderstanding! I reverted the changes. You were right, the fix for the missing parameter made those two tests pass.

lilleyse commented:
I push a commit to revert one last test: https://github.com/CesiumGS/cesium/pull/9624/commits/99f7a7b726ed592eeaa7d6a21028d0debcc81b94

Otherwise this looks good. I ran tests locally, tested `TextureEncodingTest.gltf` and `TextureEncodingTest.glb` with `ImageBitmap` enabled and disabled, and spot checked sandcastle examples.

Thanks @j9liu!
